buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -878,19 +885,24 @@ private static String hexDump(ByteBuf buffer, int fromIndex, int length) {
         }
 
         private static byte[] decodeHexDump(CharSequence hexDump, int fromIndex, int length) {
-            if (length < 0) {
+            if (length < 0 || length % 2 != 0) {
                 throw new IllegalArgumentException("length: " + length);
             }
             if (length == 0) {
                 return EmptyArrays.EMPTY_BYTES;
             }
 
-            int endIndex = fromIndex + length - 1;
             byte[] bytes = new byte[length >>> 1];
 
-            for (; fromIndex < endIndex; fromIndex += 2) {
-                bytes[fromIndex >>> 1] = (byte) ((Character.digit(hexDump.charAt(fromIndex), 16) << 4) +
-                                                  Character.digit(hexDump.charAt(fromIndex + 1), 16));
+            for (int i = 0; i < length; i += 2) {
+                int d1 = Character.digit(hexDump.charAt(fromIndex + i), 16);
```

Comment:
@fenik17 Yes, the docs explicitly states so. QueryStringDecoder.decodeHexNibble() is simple and fits our purposes better, I think.